pages/topics/new.tsx

```diff
@@ -0,0 +1,99 @@
+import { useRouter } from "next/router";
+import type { TopicProps, TopicSchema } from "$server/models/topic";
+import type { VideoTrackProps } from "$server/models/videoTrack";
+import TopicNew from "$templates/TopicNew";
+import Placeholder from "$templates/Placeholder";
+import BookNotFoundProblem from "$organisms/TopicNotFoundProblem";
+import { useAddVideoTrackAtom } from "$store/topic";
+import { updateBook, useBook } from "$utils/book";
+import { createTopic } from "$utils/topic";
+import type { Query as BookEditQuery } from "$pages/book/edit";
+import { uploadVideoTrack } from "$utils/videoTrack";
+
+export type Query = null | BookEditQuery;
+
+type NewProps = {
+  edit(topic: TopicSchema): Promise<unknown>;
+  back(): Promise<unknown>;
+  onSubmit?(topic: TopicSchema): Promise<void>;
+};
+
+function New({ edit, back, onSubmit }: NewProps) {
+  const [videoTracks, addVideoTrack] = useAddVideoTrackAtom();
+  async function handleSubmit(props: TopicProps) {
+    const topic = await createTopic(props);
+    const {
+      resource: { id: resourceId },
+    } = topic;
+
+    await Promise.all(
+      videoTracks.map((vt) => uploadVideoTrack(resourceId, vt))
+    );
+
+    await onSubmit?.(topic);
+
+    return edit(topic);
```

Comment:
```ts
    await router.replace(	
      pagesPath.book.topic.edit.$url({	
        query: { ...bookEditQuery, topicId: id },	
      })	
    );	
    return router.push(	
      pagesPath.book.edit.$url({	
        query: bookEditQuery,	
      })	
    );
```
https://github.com/npocccties/ChibiCHiLO/pull/194/commits/92e7bf382611b9739d0305fd9d91037d452e2cb8#diff-c8f1dd63659e8d8f255de00bb7b589c2ee7e9039a2f35660b0d0f45979d57bd8L39-L48

もともとここで1つ前の画面に戻っていましたが戻らなくなっていますかね。
もし作成後戻るのであれば、次の通りですかね。

```suggestion
    await edit(topic);
    return back();
```

意図通りでしょうか?

Reply:
手元で動かしながら変更した部分になりますね。

> 意図通りでしょうか?

[ブック作成](https://github.com/npocccties/ChibiCHiLO/blob/feat-cancellable-flow/pages/book/new.tsx#L11)においては、作成後ブック編集へと遷移する処理をおこなっていて、それを踏襲しました。

```ts
 async function handleSubmit(book: BookProps) {
    const { id } = await createBook(book);
    const { context }: Pick<Query, "context"> = router.query;
    await router.replace(
      pagesPath.book.edit.$url({
        query: {
          bookId: id,
          ...(context && { context }),
        },
      })
    );
  }
```

変更前のトピック作成における、作成後戻るという遷移とどちらが良いかについて、ここで議論させてもらいたいです。

変更前の遷移は、クライアント側で動画の再生時間を取得するという要件のために用意されたと想像しますが、あっていますでしょうか？

Reply:
router.replace() を加えていた意図としては、作成処理自体副作用を持つので、ブラウザーバックする必要性が無い、もしブラウザーバックするのであれば、作成画面にもどったとしても、作成前に戻れるわけではないので、編集画面に戻れるようにべき等に設計しておくほうが利用者にとってより自然ではないだろうかという仮説で追加しました。

Reply:
処理完了後、どの画面に遷移 (router.push())すべきかという判断は基本的にはFigmaにあった設計に準拠していってますかね。
どうあるべきかというと、悩ましいところですが、個人的には戻るボタンが加わったのだから、何らか処理の成功・失敗が利用者に伝わるなら、戻らずとも遷移せず実行だけでも良い気もします。

Reply:
ブックの作成画面に関しては、トピックの作成が出来ないんですよね。なのでトピック作成と同様ではないかと思います。

Reply:
> 何らか処理の成功・失敗が利用者に伝わるなら、戻らずとも遷移せず実行だけでも良い気もします。

そうですね…ブック編集もトピック編集もonSubmitでは遷移せず、成否をトースト等で表示するような挙動に変更してもよさそうですね

Reply:
> ブックの作成画面に関しては、ブック編集画面と異なっていて、トピックの作成が出来ない

もともとrouter.replace()でブック編集に遷移していたかつそれがFigma準拠になっていたのは、おっしゃるようにこの点がありましたね

Reply:
> トピック作成と同様ではないか

これは

> router.replace() を加えていた意図としては、作成処理自体副作用を持つので、ブラウザーバックする必要性が無い

ここに対応していますよね、この点に関して完全に同意です

> 個人的には戻るボタンが加わったのだから、何らか処理の成功・失敗が利用者に伝わるなら、戻らずとも遷移せず実行だけでも良い気もします

こちらのどうあるべきかについては、トピック作成のonSubmitに関する言及でしょうか？私はブック編集かトピック編集への言及として解釈しましたが

Reply:
要するにブック作成もブック編集もトピック作成もトピック編集もすべて似たような振る舞いに揃えたいですね。

Reply:
トピック作成画面とトピック編集画面は現状ほぼ同じで、同じなら同じ振る舞いにそろえたいという気持ちで、トピック編集画面が1つ前の画面に戻るのであれば、トピック作成画面も1つ前の画面に戻るのが自然かと感じます。
しかし、ブック作成画面はブック編集画面は現状異なっており、揃えるとトピック作成に不都合が出るので、ブック作成画面は戻らず、一旦、ブック編集画面に遷移後、ブック編集画面でのサブミット時に戻る、という振る舞いになるのは仕方なかろうと感じていました。

Reply:
> こちらのどうあるべきかについては、トピック作成のonSubmitに関する言及でしょうか？私はブック編集かトピック編集への言及として解釈しましたが

ブック編集、トピック編集、いずれについてもですね。

Reply:
> ブック作成はトピック作成と同様ではないかと

訂正: ブック作成はトピック作成と異なるかと

Reply:
> トピック作成画面とトピック編集画面は現状ほぼ同じで、同じなら同じ振る舞いにそろえたいという気持ち

了解です、変更前でそれを意図して用意されていましたから、そこからわざわざ遷移を変更する必要性はないと理解しました！